### PR TITLE
Default field limit

### PR DESCRIPTION
**Pull request by afrasier**
Added default field limiting and nested_serializer meta
parameterization to Awards/Summary

### CODE REVIEW

**Comment by afrasier** [comment]
This approach will allow us, in the future, to finish nested serializer field limiting on demand by parsing incoming field lists and customizing child serializer context.

File: usaspending_api/common/serializers.py [hunk 1]
```diff
@@ -12,6 +12,20 @@ def __init__(self, *args, **kwargs):
...
+        try:
+            # Initialize the child serializers
+            children = self.Meta.nested_serializers
+            for field in children.keys():
+                child_args = {
+                    **children[field].get("kwargs", {}),
+                    "context": self.context
+                }
+                self.fields[field] = children[field]["class"](**child_args)
+        except:
```
Comment (afrasier) [inline usaspending_api/common/serializers.py]: I think it should be fine as the list serializer will have a list of LimitableSerializers, we'll just need to override many_init to add a line to pass the context to the children.

Yeah, we should catch specifically AttributeNotFound exceptions here. I will update that.

**Comment by bsweger** [comment]
@afrasier Apologies for not asking before--do we need tests for this?

**Comment by afrasier** [comment]
> Most importantly, piid, uri, and fain should be in their own Award ID object.

This can be done either in the object (requiring yet-another-model change and a big data migration) or in the serializer (which is free). Either case front-end will require some updates. The only issue with the latter solution is when filtering, you'd still have to filter on `award__piid` rather than `award__award_id__piid` which may get confusing. Hence why I just left them at the root level for this PR.

> place_of_performance

This didn't exist when the spec was created, but seemed important enough to include as well

> awarding agency: create/update dates
> financialassistance award set: create/update dates
> procurement award set: create/update dates

Oversight, I'll remove it.

> location: country code

A lot of the front end lookups use country code rather than name, so I left this in there as it is good to have in addition to the name

> Lastly, are `procurement_set `and `financialassistanceaward_set `analagous to `transaction_list `in the spec? I'm not sure what the difference is supposed to be.

Essentially. I believe the want here was to unify procurements and financial assistance into the same return array, but this causes issues because they are not really interchangeable. It can be done in the serializer so the name is always transaction_list, but that comes at the cost of increasing the requirements on the end user to know how to request a specific filter. (Much like in the earlier example about the award id object)

**Comment by afrasier** [comment]
@bsweger Tests were updated. (Testing via test_endpoints.py)

File: usaspending_api/common/serializers.py [hunk 2]
```diff
@@ -22,7 +22,7 @@ def __init__(self, *args, **kwargs):
                     "context": self.context
                 }
                 self.fields[field] = children[field]["class"](**child_args)
-        except:
+        except AttributeError:
```
Comment (bsweger) [inline usaspending_api/common/serializers.py]: Thanks!

**Comment by bsweger** [comment]
@afrasier thanks for the changes!

This may not close out the story, but I'm pro getting this round of changes into the code base sooner rather than later, and I approve the PR. Someone from the front-end team ( @kevinli-work, @mab6bf, or @ejgullo  ) should give the official approval, however, to make sure we haven't hidden by default any of the fields they're using.

**Comment by bsweger** [comment]
Posting sample `awards/summary` response objects for front-end review:

**Contract**

```json
{
  "total_metadata": {
    "count": 9627
  },
  "page_metadata": {
    "num_pages": 9627,
    "page_number": 1,
    "count": 1
  },
  "results": [
    {
      "id": 47963,
      "type": "C",
      "type_description": "Delivery Order",
      "piid": "1380",
      "fain": null,
      "uri": null,
      "total_obligation": "2074.80",
      "total_outlay": null,
      "date_signed": "2013-10-21",
      "description": null,
      "period_of_performance_start_date": "2013-10-21",
      "period_of_performance_current_end_date": "2013-10-21",
      "awarding_agency": {
        "toptier_agency": {
          "cgac_code": "097",
          "fpds_code": "9700",
          "name": "DEPT OF DEFENSE"
        },
        "subtier_agency": {
          "subtier_code": "97AS",
          "name": "DEFENSE LOGISTICS AGENCY"
        },
        "office_agency": null
      },
      "funding_agency": {
        "toptier_agency": {
          "cgac_code": "097",
          "fpds_code": "9700",
          "name": "DEPT OF DEFENSE"
        },
        "subtier_agency": {
          "subtier_code": "97AS",
          "name": "DEFENSE LOGISTICS AGENCY"
        },
        "office_agency": null
      },
      "recipient": {
        "legal_entity_id": 927146409,
        "ultimate_parent_legal_entity_id": null,
        "recipient_name": "TRI-STATE GOVERNMENT SERVICES, INC.",
        "business_types": null,
        "location": {
          "location_country_name": null,
          "location_state_code": "TN",
          "location_state_name": null,
          "location_city_name": "CHATTANOOGA",
          "location_address_line1": "4762 HIGHWAY 58 STE H",
          "location_address_line2": "",
          "location_address_line3": "",
          "location_zip5": "37416",
          "location_foreign_postal_code": null,
          "location_foreign_province": null,
          "location_foreign_city_name": null,
          "location_country_code": "USA"
        }
      },
      "place_of_performance": 58,
      "financialassistanceaward_set": [],
      "procurement_set": [
        {
          "action_date": "2013-10-21",
          "federal_action_obligation": "1037.40",
          "modification_number": "1",
          "award_description": null,
          "cost_or_pricing_data": "N",
          "naics": "NONE",
          "naics_description": null,
          "idv_type": null,
          "product_or_service_code": "S222",
          "update_date": "2017-01-13T19:57:11.164361Z"
        },
        {
          "action_date": "2013-10-21",
          "federal_action_obligation": "1037.40",
          "modification_number": "1",
          "award_description": null,
          "cost_or_pricing_data": "N",
          "naics": "NONE",
          "naics_description": null,
          "idv_type": null,
          "product_or_service_code": "S222",
          "update_date": "2017-01-13T19:52:54.870034Z"
        }
      ]
    }
  ]
}
```

**Financial Assistance**

```json
{
  "page_metadata": {
    "num_pages": 3,
    "count": 2,
    "page_number": 1
  },
  "total_metadata": {
    "count": 5
  },
  "results": [
    {
      "id": 21389,
      "type": "05",
      "type_description": "Cooperative Agreement",
      "piid": null,
      "fain": "SBAHQ16B0017",
      "uri": null,
      "total_obligation": "400000.00",
      "total_outlay": null,
      "date_signed": "2016-09-16",
      "description": "SMALL BUSINESS DEVELOPMENT CENTERS",
      "period_of_performance_start_date": "2016-09-16",
      "period_of_performance_current_end_date": "2016-09-16",
      "awarding_agency": {
        "id": 1487,
        "create_date": "2017-01-12T19:56:29.014000Z",
        "update_date": "2017-01-12T19:56:29.014000Z",
        "toptier_agency": {
          "cgac_code": "073",
          "fpds_code": "7300",
          "name": "SMALL BUSINESS ADMINISTRATION"
        },
        "subtier_agency": {
          "subtier_code": "7300",
          "name": "SMALL BUSINESS ADMINISTRATION"
        },
        "office_agency": null
      },
      "funding_agency": {
        "id": 1487,
        "create_date": "2017-01-12T19:56:29.014000Z",
        "update_date": "2017-01-12T19:56:29.014000Z",
        "toptier_agency": {
          "cgac_code": "073",
          "fpds_code": "7300",
          "name": "SMALL BUSINESS ADMINISTRATION"
        },
        "subtier_agency": {
          "subtier_code": "7300",
          "name": "SMALL BUSINESS ADMINISTRATION"
        },
        "office_agency": null
      },
      "recipient": {
        "legal_entity_id": 782279541,
        "ultimate_parent_legal_entity_id": null,
        "recipient_name": "UNIVERSITY OF LOUISIANA AT MONROE",
        "business_types": "H",
        "location": {
          "location_country_name": "UNITED STATES",
          "location_state_code": "LA",
          "location_state_name": "Louisiana",
          "location_city_name": "Monroe",
          "location_address_line1": "700 University Ave",
          "location_address_line2": null,
          "location_address_line3": null,
          "location_zip5": "71209",
          "location_foreign_postal_code": null,
          "location_foreign_province": null,
          "location_foreign_city_name": null,
          "location_country_code": "USA"
        }
      },
      "place_of_performance": 5617,
      "procurement_set": [],
      "financialassistanceaward_set": [
        {
          "action_date": "2016-09-16",
          "federal_action_obligation": "150209.00",
          "modification_number": "0003:560400DB",
          "award_description": "SMALL BUSINESS DEVELOPMENT CENTERS",
          "cfda_number": "59.037",
          "cfda_title": "Small Business Development Centers",
          "face_value_loan_guarantee": null,
          "original_loan_subsidy_cost": null,
          "assistance_type": "05",
          "update_date": "2017-01-13T08:38:43.618327Z"
        },
        {
          "action_date": "2016-09-16",
          "federal_action_obligation": "249791.00",
          "modification_number": "0003:670400DB",
          "award_description": "SMALL BUSINESS DEVELOPMENT CENTERS",
          "cfda_number": "59.037",
          "cfda_title": "Small Business Development Centers",
          "face_value_loan_guarantee": null,
          "original_loan_subsidy_cost": null,
          "assistance_type": "05",
          "update_date": "2017-01-13T08:38:41.765328Z"
        }
      ]
    },
    {
      "id": 22044,
      "type": "05",
      "type_description": "Cooperative Agreement",
      "piid": null,
      "fain": "SBAHQ16V0028",
      "uri": null,
      "total_obligation": "249997.23",
      "total_outlay": null,
      "date_signed": "2016-09-19",
      "description": "FY 2016 B2BRP",
      "period_of_performance_start_date": "2016-09-19",
      "period_of_performance_current_end_date": "2016-09-19",
      "awarding_agency": {
        "id": 1487,
        "create_date": "2017-01-12T19:56:29.014000Z",
        "update_date": "2017-01-12T19:56:29.014000Z",
        "toptier_agency": {
          "cgac_code": "073",
          "fpds_code": "7300",
          "name": "SMALL BUSINESS ADMINISTRATION"
        },
        "subtier_agency": {
          "subtier_code": "7300",
          "name": "SMALL BUSINESS ADMINISTRATION"
        },
        "office_agency": null
      },
      "funding_agency": {
        "id": 1487,
        "create_date": "2017-01-12T19:56:29.014000Z",
        "update_date": "2017-01-12T19:56:29.014000Z",
        "toptier_agency": {
          "cgac_code": "073",
          "fpds_code": "7300",
          "name": "SMALL BUSINESS ADMINISTRATION"
        },
        "subtier_agency": {
          "subtier_code": "7300",
          "name": "SMALL BUSINESS ADMINISTRATION"
        },
        "office_agency": null
      },
      "recipient": {
        "legal_entity_id": 927146407,
        "ultimate_parent_legal_entity_id": null,
        "recipient_name": "ASSOCIATION OF SMALL BUSINESS DEVELOPMENT CEN",
        "business_types": "M",
        "location": {
          "location_country_name": "UNITED STATES",
          "location_state_code": "VA",
          "location_state_name": "Virginia",
          "location_city_name": "Burke",
          "location_address_line1": "8990 Burke Lake Rd",
          "location_address_line2": null,
          "location_address_line3": null,
          "location_zip5": "22015",
          "location_foreign_postal_code": null,
          "location_foreign_province": null,
          "location_foreign_city_name": null,
          "location_country_code": "USA"
        }
      },
      "place_of_performance": 5611,
      "procurement_set": [],
      "financialassistanceaward_set": [
        {
          "action_date": "2016-09-19",
          "federal_action_obligation": "17853.97",
          "modification_number": "BASE:560400DB",
          "award_description": "FY 2016 B2BRP",
          "cfda_number": "59.044",
          "cfda_title": "Veterans Business Development",
          "face_value_loan_guarantee": null,
          "original_loan_subsidy_cost": null,
          "assistance_type": "05",
          "update_date": "2017-01-13T08:38:30.047283Z"
        },
        {
          "action_date": "2016-09-19",
          "federal_action_obligation": "232143.26",
          "modification_number": "BASE:670400DB",
          "award_description": "FY 2016 B2BRP",
          "cfda_number": "59.044",
          "cfda_title": "Veterans Business Development",
          "face_value_loan_guarantee": null,
          "original_loan_subsidy_cost": null,
          "assistance_type": "05",
          "update_date": "2017-01-13T08:38:28.187788Z"
        }
      ]
    }
  ]
}
```

**Comment by kevinli-work** [comment]
Fields look fine. With the exception of the nested fields, frontend is explicitly requesting fields anyway. Nested fields looks fine, though having the fields documented would be nicer for review in the future than reading Python 😃

**Pull request merged**